x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java

```diff
@@ -122,37 +141,66 @@ private void handleFailure(Exception e, ActionListener<GetPipelineResponse> list
         }
     }
 
-    private void handleSearchResponse(
+    private void handleFilteringSearchResponse(
         SearchResponse searchResponse,
         Map<String, BytesReference> pipelineSources,
+        Set<String> explicitPipelineIds,
+        Set<String> wildcardPipelineIdPatterns,
+        int maxNumHitsToFetch,
         Consumer<SearchResponse> clearScroll,
         ActionListener<GetPipelineResponse> listener
     ) {
+        maxNumHitsToFetch += searchResponse.getHits().getHits().length;
         for (SearchHit hit : searchResponse.getHits().getHits()) {
-            pipelineSources.put(hit.getId(), hit.getSourceRef());
+            if (explicitPipelineIds.isEmpty() && wildcardPipelineIdPatterns.isEmpty()) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+
+            // take if ID is in request IDs set
+            if (explicitPipelineIds.contains(hit.getId())) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+            // take if id matches request wildcard pattern
+            for (String wildcardPipelineIdPattern : wildcardPipelineIdPatterns) {
+                if (Pattern.matches(wildcardPipelineIdPattern, hit.getId())) {
+                    pipelineSources.put(hit.getId(), hit.getSourceRef());
+                    break;
+                }
+            }
```

Comment:
Thanks @kaisecheng for the details,
If you are worrying about performance, as my understanding it will not matter since JIT optimizer takes care.
From clean code perspective, it really varies from person to person. My approach is based on lambda where I use `filter` and `map` in plain loop as well. Where I am filtering by continuing and mapping at the end. If we have strict code formatting I am happy to change.